File: frontend/src/data/design/elevator.js
// designDetails — elevator
// Single source of truth for this module. One file per module: duplicate keys in a
// shared object literal previously let JavaScript silently discard the richer entry.

export default {
  title: 'Elevator Control System — Design Details',
  tldr: [
    'Multi-elevator control system dispatching a 4-car bank across a 10-floor building',
    'Two swappable dispatch strategies — LOOK/SCAN direction-aware scoring and a raw Nearest-Car baseline — resolved through an EnumMap-backed factory',
    'A declared elevator state machine (IDLE / MOVING_UP / MOVING_DOWN / DOOR_OPEN / MAINTENANCE) with a legal-transition table enforced on every state change, not just tracked by an unguarded enum field',
    'Per-elevator ReentrantLock plus a controller-wide dispatch lock so concurrent floor calls are assigned exactly once, with no double-booked or lost requests',
    'Isolated /api/elevator/sim/* sandbox so the interactive demo can never corrupt the real elevator bank'
  ],
  requirements: [
    'Multi-elevator dispatch across N floors (4 elevators, 10 floors)',
    'External floor call buttons that already know both the pickup floor and the intended destination',
    'Internal elevator destination buttons for a rider who wants a different floor than announced',
    'Optimal elevator selection based on proximity, current direction, and remaining capacity, with a choice of dispatch algorithm',
    'A real door-open/close lifecycle that the JSON contract reports honestly (not collapsed into a generic "stopped" status)',
    'Pulling a car into maintenance mid-route without losing or double-assigning its queued stops'
  ],
  entities: [
    {
      name: 'ElevatorControllerService',
      description: 'Owns both the real elevator bank and an isolated /sim/* sandbox. Dispatches, steps the simulation clock, and enforces floor-range and elevator-existence validation before anything touches an Elevator.',
      fields: [
        { name: 'activePolicy', type: 'DispatchPolicy', description: 'Which strategy the factory resolves for every dispatch call right now.' },
        { name: 'controllerLock', type: 'ReentrantLock', description: 'Serializes dispatch decisions so two simultaneous calls can never race for the same car.' },
        { name: 'pendingExternalRequests', type: 'Queue<Request>', description: 'Calls that found no eligible car at assignment time; drained on every tick and whenever a car returns from maintenance.' }
      ],
      methods: [
        { name: 'handleExternalRequest(sourceFloor, destinationFloor)', returns: 'Request', description: 'Validates both floors, dispatches via the active strategy, and queues both stops on the winning car in one step.' },
        { name: 'handleInternalRequest(elevatorId, destinationFloor)', returns: 'void', description: 'Adds a destination call to a specific car already in service — rejects a MAINTENANCE car.' },
        { name: 'setElevatorMaintenance(elevatorId, maintenance)', returns: 'void', description: 'Guarded MAINTENANCE transition; reassigns every orphaned stop as a fresh external request.' },
        { name: 'stepSimulation() / simStep()', returns: 'List<Elevator>', description: 'Advances every non-maintenance car by one floor or one door-timer tick.' }
      ]
    },
    {
      name: 'Elevator',
      description: 'Thread-safe per-car state: current floor, direction, occupancy, and two ConcurrentSkipListSets tracking queued up/down stops. Not converted to Lombok — like trafficsignal.model.Intersection and atm.model.Account, a lock-holding entity with real invariants stays hand-written rather than accepting generated setters that would bypass the guard.',
      fields: [
        { name: 'state', type: 'ElevatorState', description: 'Mutated only through transitionTo(), never assigned directly outside test fixtures.' },
        { name: 'elevatorLock', type: 'ReentrantLock', description: 'Guards every read-then-write on this car\'s stops, occupancy and state.' }
      ],
      methods: [
        { name: 'transitionTo(target)', returns: 'void', description: 'Validates target against ElevatorLifecycleStates.of(state).allowedNext() and throws IllegalElevatorStateTransitionException for an illegal jump.' },
        { name: 'addStop(floor) / removeStop(floor)', returns: 'void', description: 'Maintains the ascending up-stop and descending down-stop sets.' },
        { name: 'boardPassenger() / deboardPassenger(n)', returns: 'boolean / void', description: 'Atomic occupancy changes bounded by capacity.' }
      ]
    },
    {
      name: 'ElevatorDispatchStrategyFactory',
      description: 'Resolves DispatchPolicy to its ElevatorDispatchStrategy via an EnumMap built once in the constructor — the same shape as inventory.strategy.ReorderStrategyFactory.',
      fields: [],
      methods: [
        { name: 'forPolicy(policy)', returns: 'ElevatorDispatchStrategy', description: 'One O(1) lookup, no branching at call sites.' }
      ]
    }
  ],
  designPatterns: [
    {
      name: 'Strategy Pattern + Factory',
      used: true,
      explanation: 'ElevatorDispatchStrategy has two implementations — LookScanDispatchStrategy (distance + direction-penalty scoring with 3-tier tie-breaking) and NearestCarDispatchStrategy (pure closest-car, ignoring direction of travel entirely) — resolved by ElevatorDispatchStrategyFactory\'s EnumMap<DispatchPolicy, ElevatorDispatchStrategy>. Switching policy at runtime (POST /api/elevator/policy) changes which car gets picked without touching ElevatorControllerService.'
    },
    {
      name: 'State Pattern',
      used: true,
      explanation: 'com.lld.elevator.state — one singleton per ElevatorState (IdleState, MovingUpState, MovingDownState, DoorOpenState, MaintenanceState), each declaring its own Set<ElevatorState> allowedNext(), the same class-per-state shape as taskmanagement.state.TaskState. Elevator#transitionTo is the single enforcement point; an illegal jump (e.g. MOVING_UP straight to IDLE, skipping the door-open phase) throws IllegalElevatorStateTransitionException instead of silently overwriting the field.'
    },
    {
      name: 'Observer Pattern',
      used: true,
      explanation: 'ElevatorNotifier fans state-change/floor-reached/door-change events out to every registered ElevatorObserver — LoggingElevatorObserver (audit trail to the app log) and InMemoryElevatorEventObserver (bounded in-memory ring buffer), both Spring-managed beans injected into the live notifier. The isolated /sim/* sandbox deliberately does not route through this notifier — it keeps its own SimEvent log — so a replayed demo can never appear in the real telemetry stream.'
    }
  ],
  principles: [
    { name: 'Single Responsibility', description: 'Elevator owns car-local state and stop bookkeeping; ElevatorControllerService owns dispatch, ticking and maintenance orchestration; ElevatorDispatchStrategyFactory owns policy resolution.' },
    { name: 'Open/Closed', description: 'Adding a third dispatch policy means one new ElevatorDispatchStrategy implementation, one enum constant, one factory entry — zero changes to ElevatorControllerService.' },
    { name: 'Encapsulation', description: 'Elevator\'s up/down stop sets and state field are private; every mutation goes through a method that maintains the class invariants (bounded occupancy, legal transitions only).' }
  ],
  oopConcepts: [
    { name: 'Polymorphism', description: 'ElevatorControllerService calls strategy.selectOptimalElevator(...) and state.canTransitionTo(...) without knowing which concrete implementation is behind either interface.' },
    { name: 'Encapsulation', description: 'The elevator\'s internal floor queues (upStops/downStops) are only ever mutated via addStop/removeStop, never exposed as a mutable reference for a caller to corrupt directly.' }
  ],
  extensibility: [
    { area: 'Express elevators', description: 'A third DispatchPolicy that only stops at even floors above a threshold — one new strategy implementation.', difficulty: 'Medium' },
    { area: 'Weight-based capacity', description: 'Swap the passenger-count capacity check in ElevatorDispatchStrategy for a weight sensor input.', difficulty: 'Medium' },
    { area: 'Multi-bank buildings', description: 'Partition the elevator list by bank before handing it to the strategy — the strategy interface does not change.', difficulty: 'Hard' }
  ],
  tradeoffs: [
    'LOOK/SCAN over pure FCFS to minimize total travel distance and average wait time — the direction-penalty scoring in LookScanDispatchStrategy strongly prefers a car already headed toward the call over an idle-but-farther one.',
    'Dispatch takes both the source and destination floor in one call rather than a strict two-phase "press the hall button, then press a floor button once inside" flow — a documented simplification that also fixed a real bug: the earlier two-phase design queued a placeholder stop one floor past the pickup that was never removed, and completeMatchingRequests could never match a request against its real destination floor.',
    'The reassignment path for a car pulled into maintenance mid-route only knows a bare floor number was pending, not whether it was the original pickup or destination — so the requeued request\'s displayed destination is a same-direction placeholder. A documented limitation of that specific recovery path, not a resurfacing of the two-phase bug above.',
    'Background @Scheduled ticking (1.5s) drives the real elevator bank so the UI can simply poll GET /elevators; the isolated sim sandbox is stepped explicitly by the caller instead, so a demo replay is fully deterministic and pausable.'
  ]
};

File: backend/src/main/java/com/lld/elevator/model/Elevator.java
package com.lld.elevator.model;

import com.lld.elevator.exception.IllegalElevatorStateTransitionException;
import com.lld.elevator.state.ElevatorLifecycleStates;

import java.util.*;
import java.util.concurrent.ConcurrentSkipListSet;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.ReentrantLock;

public class Elevator {

    private long id;
    private String name;
    private volatile int currentFloor;
    private volatile int destinationFloor;
    private volatile Direction direction;
    private volatile ElevatorState state;
    private int capacity;
    private final AtomicInteger currentOccupancy = new AtomicInteger(0);

    // Concurrent skip list sets for LOOK/SCAN algorithm stops
    private final Set<Integer> upStops = new ConcurrentSkipListSet<>();
    private final Set<Integer> downStops = new ConcurrentSkipListSet<>(Collections.reverseOrder());

    private final ReentrantLock elevatorLock = new ReentrantLock(true);

    public Elevator() {
        this.direction = Direction.IDLE;
        this.state = ElevatorState.IDLE;
        this.currentFloor = 1;
        this.destinationFloor = 1;
        this.capacity = 8;
    }

    public Elevator(long id, String name, int capacity, int currentFloor) {
        this();
        this.id = id;
        this.name = name;
        this.capacity = capacity;
        this.currentFloor = currentFloor;
        this.destinationFloor = currentFloor;
    }

    public Elevator(int id, String name, int capacity, int currentFloor) {
        this((long) id, name, capacity, currentFloor);
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getCurrentFloor() {
        return currentFloor;
    }

    public void setCurrentFloor(int currentFloor) {
        this.currentFloor = currentFloor;
    }

    public int getDestinationFloor() {
        return destinationFloor;
    }

    public void setDestinationFloor(int destinationFloor) {
        this.destinationFloor = destinationFloor;
    }

    public Direction getDirection() {
        return direction;
    }

    public void setDirection(Direction direction) {
        this.direction = direction;
    }

    public ElevatorState getState() {
        return state;
    }

    /**
     * Raw setter — bypasses the legal-transition table. Reserved for constructing test fixtures
     * and seeding initial state; production call sites should use {@link #transitionTo(ElevatorState)}.
     */
    public void setState(ElevatorState state) {
        this.state = state;
    }

    /**
     * Guarded transition: validates {@code target} against this elevator's current
     * {@link com.lld.elevator.state.ElevatorLifecycleState#allowedNext()} before applying it.
     * This is the one place production code should ever change {@link #state}.
     *
     * @throws IllegalElevatorStateTransitionException if {@code target} is not a legal next state
     */
    public void transitionTo(ElevatorState target) {
        if (!ElevatorLifecycleStates.of(this.state).canTransitionTo(target)) {
            throw new IllegalElevatorStateTransitionException(
                    "Elevator " + id + " (" + name + ") cannot transition from " + this.state + " to " + target);
        }
        this.state = target;
    }

    /**
     * The coarse-grained status the frontend renders. Previously collapsed {@code DOOR_OPEN} into
     * {@code STOPPED}, which meant the JSON contract could never actually tell a caller doors
     * were open — the frontend had to fake it with a client-side timer guess. Now that
     * {@link ElevatorState#DOOR_OPEN} is a real, guarded state, it is reported honestly.
     */
    public ElevatorStatus getStatus() {
        if (state == ElevatorState.MAINTENANCE) return ElevatorStatus.OUT_OF_ORDER;
        if (state == ElevatorState.MOVING_UP || state == ElevatorState.MOVING_DOWN) return ElevatorStatus.MOVING;
        if (state == ElevatorState.DOOR_OPEN) return ElevatorStatus.DOOR_OPEN;
        return ElevatorStatus.STOPPED;
    }

    public void setStatus(ElevatorStatus status) {
        if (status == ElevatorStatus.OUT_OF_ORDER) {
            this.state = ElevatorState.MAINTENANCE;
        } else if (status == ElevatorStatus.MOVING) {
            this.state = (direction == Direction.DOWN) ? ElevatorState.MOVING_DOWN : ElevatorState.MOVING_UP;
        } else {
            this.state = ElevatorState.IDLE;
        }
    }

    public int getCapacity() {
        return capacity;
    }

    public void setCapacity(int capacity) {
        this.capacity = capacity;
    }

    public int getCurrentOccupancy() {
        return currentOccupancy.get();
    }

    public int getCurrentLoad() {
        return currentOccupancy.get();
    }

    public void setCurrentLoad(int load) {
        this.currentOccupancy.set(load);
    }

    public boolean boardPassenger() {
        elevatorLock.lock();
        try {
            if (currentOccupancy.get() < capacity) {
                currentOccupancy.incrementAndGet();
                return true;
            }
            return false;
        } finally {
            elevatorLock.unlock();
        }
    }

    public void deboardPassenger(int count) {
        elevatorLock.lock();
        try {
            int cur = currentOccupancy.get();
            currentOccupancy.set(Math.max(0, cur - count));
        } finally {
            elevatorLock.unlock();
        }
    }

    public boolean isFull() {
        return currentOccupancy.get() >= capacity;
    }

    public void addDestination(int floor) {
        addStop(floor);
    }

    public void addStop(int floor) {
        elevatorLock.lock();
        try {
            if (floor > currentFloor) {
                upStops.add(floor);
            } else if (floor < currentFloor) {
                downStops.add(floor);
            } else {
                if (direction == Direction.DOWN) {
                    downStops.add(floor);
                } else {
                    upStops.add(floor);
                }
            }
        } finally {
            elevatorLock.unlock();
        }
    }

    public void removeStop(int floor) {
        elevatorLock.lock();
        try {
            upStops.remove(floor);
            downStops.remove(floor);
        } finally {
            elevatorLock.unlock();
        }
    }

    public List<Integer> getPendingFloors() {
        elevatorLock.lock();
        try {
            List<Integer> list = new ArrayList<>();
            list.addAll(upStops);
            list.addAll(downStops);
            return list;
        } finally {
            elevatorLock.unlock();
        }
    }

    public Set<Integer> getUpStops() {
        return upStops;
    }

    public Set<Integer> getDownStops() {
        return downStops;
    }

    public ReentrantLock getLock() {
        return elevatorLock;
    }

    public ElevatorSnapshot createSnapshot() {
        return new ElevatorSnapshot(
            id, name, currentFloor, state, direction,
            currentOccupancy.get(), capacity,
            new ArrayList<>(upStops), new ArrayList<>(downStops)
        );
    }
}
